### ISSUE TITLE
Allow custom user authentication

### ISSUE THREAD

**Issue by hanak** [wduckitt/React-Automation-Studio#13]
This is a proposal for a potential future enhancement. The purpose of this ticket (issue) is to identify whether this is desired functionality and how it should be handled.

**Problem**
pvServer includes user authentication and permission validation. Authentication uses a JSON file with encrypted content in order to define users. Other sources of user authentication are not supported.

**Potential Design**
Custom authentication solution uses aggregation in order to supply necessary functionality. Text below is just a sketch of a design as I assume it misses some details:
* Split `userAuthentication/authenticate.py` into two modules: (1) permission/access validation (`permissions.py`) and (2) authentication (`authentication.py`)
* Add a function `set_authentication()` to `pvServer.py`. This function will accept a reference to an instance of a class which handles custom authentication. If no custom authentication is supplied, `pvServer.py` will load `authentication.py` (the part (2) of `userAuthentication/authenticate.py`) instead.
* Add a function `run()` to pvServer.py which consist of code after `if __name__ == '__main__':`. Update `pvServer.py` with following:
```
if __name__ == '__main__':
    run()
```

Expected use when supplying custom authentication:
```
import pvServer
pvServer.set_authentication(my_authentication)
pvServer.run()
```

**Comment by wduckitt** [comment]
Hi

Yes the plan is for a web based admin. The goal was first to replace the
control system studio functionality. We added in the alarm handler and load
save, which uses the mongo database. Once the archiver viewer is done we
will move onto the admin side. I think we will change the authentication to
be based on the mongo db, but leave  in a hook to do authentication another
way. Authorisation would though need to be based on the what ever is
defined in the mongodb.

We can  have a zoom meeting and thrash it out in a few weeks time.

Regards,

William

**Comment by hanak** [comment]
Hello,

Thank you for your response. I think this sounds interesting and it will make administration easier. If the solution allowed changing UAG without restart, that would be great. And, if the hook is kept, one can use a different server for authentication purposes. Thus, it all sounds good.

Thank you.

Sincerely,
Ivo

**Comment by wduckitt** [comment]
Hi Ivo

Im going to statr looking into this again

What authentication process would prefer to use at your lab?

We use active directory

**Comment by hanak** [comment]
Hello,

Thank you for your response. I need to confirm that I can share this information first. If so, I will share through email found in the `README.md`. Beside the production solution, I use the json file for local development as I do not have permissions to add real users. It would be great if the solution allowed inserting a plugin which could authenticate users. Then, you would not need to spend time on integrating with various environments.

Thank you.

Ivo.

**Comment by hanak** [comment]
Hello,

I am sorry for late response. When I asked around I did not get a clear confirmation. Thus, I am inclining towards using a plugin/hook to implement custom authentication. It seems to me that such an approach will help React Automation Studio to offer working approach to customization as I assume there are more methods out there. As you have mentioned you will be working on a new solution to authentication and authorization. I can contribute by either implementing or testing a plugin/hook to the authentication. Would you find this acceptable?

Thank you.

Ivo.

**Comment by wduckitt** [comment]
Flask has verious methods, but I'll need to manage the integration method.
It have started looking into by I doubt it will be ready this year due to
other commitments

**Comment by wduckitt** [comment]
It will come with the web based admin

**Comment by darcato** [comment]
Hello, we are also interested in a custom authentication mechanism. We are planning to use OpenID Connect to authenticate, authorize and manage the user externally, in a centralized way. It would be nice to have a way to integrate it with this application.

**Comment by wduckitt** [comment]
Hi All

I haven't had time in the last while to work on it. So I need assistance.

In a nut shell flask supports various authentication processes.

The trick is to make the system support many of them without over
complicating it.

If you guys are willing to help I propose this:

Independent of the current implementation of RAS we must determine within
the Python/Flask  environment how to do it.

Perhaps someone can do an analysis of how to do it and then we setup a
basic project that  uses the latest flask (yes I need to update it)  and a
simple react app that just has login and user info and then does the
specific authentication based on env parameters.

We can then test this app at our facilities and then make the changes to
RAS.

What do you guys think?

Regards,
William

**Comment by wduckitt** [comment]
Can you guys send me an email at rasadmin@tlabs.ac.za if you are interested
in collaborating on it, and then I'll setup a Zoom meeting.

**Comment by wduckitt** [comment]
Im closing the issue as V3.0.0  goes a long way to solve this. Please repoen if you any other suggestions
